open_mastr/xml_download/utils_write_to_database.py

```diff
@@ -392,3 +293,51 @@ def find_nearest_brackets(xml_string: str, position: int) -> tuple[int, int]:
             continue
 
     raise Error("An error occured when parsing the xml file. Maybe it is corrupted?")
+
+
+def process_table_before_insertion(
+    df: pd.DataFrame,
+    xml_table_name: str,
+    zipped_xml_file_path: str,
+    bulk_download_date: str,
+    bulk_cleansing: bool,
+) -> pd.DataFrame:
+    df = add_zero_as_first_character_for_too_short_string(df)
+    df = change_column_names_to_orm_format(df, xml_table_name)
+
+    # Add Column that refers to the source of the data
+    df["DatenQuelle"] = "bulk"
+    df["DatumDownload"] = bulk_download_date
+
+    if bulk_cleansing:
+        df = cleanse_bulk_data(df, zipped_xml_file_path)
+    return df
+
+
+def add_table_to_database(
+    df: pd.DataFrame,
+    xml_table_name: str,
+    sql_table_name: str,
+    engine: sqlalchemy.engine.Engine,
+) -> None:
+    column_list = df.columns.tolist()
+    add_missing_columns_to_table(engine, xml_table_name, column_list)
+
+    # Convert NaNs to None.
+    df = df.where(pd.notnull(df), None)
+
+    # Convert date columns to strings. Dates are not supported directly by SQLite.
+    df = cast_date_columns_to_string(xml_table_name, df)
+
+    # Create SQL statement for bulk insert. ON CONFLICT DO NOTHING prevents duplicates.
+    insert_stmt = f"INSERT INTO {sql_table_name} ({','.join(column_list)}) VALUES ({','.join(['?' for _ in column_list])}) ON CONFLICT DO NOTHING"
+
+    for _ in range(10000):
+        try:
+            with engine.connect() as con:
+                with con.begin():
+                    con.connection.executemany(insert_stmt, df.to_numpy())
```

Comment:
One mistake that is probably my fault, since we never really talked about that: The code needs to work for two different database types: `sqlite` and `postgres`. It seems like postgres does not provide a `con.connection.executemany`. 
I assume that maybe 90% of users will have a sqlite database. Hence I see two possible solutions:
1. There is a way to change this code so that it works for postgres as well but does not decrease the parsing speed for sqlite.
2. You create two functions, one that is called when the engine is of type sqlite to use the fast code, and one that is called if the engine is of any other type. Then we use the old pandas.to_sql function and hope that pandas can handle all the other engine types.

Reply:
Do you agree @AlexandraImbrisca or do you see another way to go?

Reply:
While I'm trying to set up PostgreSQL, is [this](https://github.com/OpenEnergyPlatform/open-MaStR/blob/bd44ce26986a0c41b25250951a4a9a391723d690/docs/advanced.md?plain=1#L25) the usage you are referring to?

Reply:
Yes, exactly. Let's stick with option 2 for now. I think getting this PR and the PR about parallelization merged has a higher priority right now.

Reply:
Sounds great, thank you! I updated the code to switch between the 2 implementations of the ```add_table_to_database``` function based on the dialect. The rest should remain the same since it doesn't rely on the database.

I parametrized the test for add_table_to_database and the functions produce the same output. Please take another look whenever you have time! I'll start the parallelization PR